pkg/skaffold/server/server.go

```diff
@@ -96,18 +96,25 @@ func SetAutoSyncCallback(callback func(bool)) {
 // It returns a shutdown callback for tearing down the grpc server,
 // which the runner is responsible for calling.
 func Initialize(opts config.SkaffoldOptions) (func() error, error) {
-	if !opts.EnableRPC || opts.RPCPort == -1 {
-		return func() error { return nil }, nil
+	emptyCallback := func() error { return nil }
+	if !opts.EnableRPC && opts.RPCPort.Value() == nil && opts.RPCHTTPPort.Value() == nil {
+		logrus.Debug("skaffold API not starting as it's not requested")
+		return emptyCallback, nil
 	}
 
-	var usedPorts util.PortSet
-
-	grpcCallback, rpcPort, err := newGRPCServer(opts.RPCPort, &usedPorts)
+	preferredGRPCPort := 0 // bind to an available port atomically
+	if opts.RPCPort.Value() != nil {
+		preferredGRPCPort = *opts.RPCPort.Value()
+	}
+	grpcCallback, grpcPort, err := newGRPCServer(preferredGRPCPort)
 	if err != nil {
 		return grpcCallback, fmt.Errorf("starting gRPC server: %w", err)
 	}
+	httpCallback := emptyCallback
 
-	httpCallback, err := newHTTPServer(opts.RPCHTTPPort, rpcPort, &usedPorts)
+	if opts.RPCHTTPPort.Value() != nil {
+		httpCallback, err = newHTTPServer(*opts.RPCHTTPPort.Value(), grpcPort)
+	}
```

Comment:
If `--enable-rpc` was specified alone, then we need to implicitly work as if the user specified `--rpc-port=50051 --rpc-http-port=50052`.

Since you're disabling the `--enable-rpc` default for `dev` and `debug`, failing if the ports are unavailable is the right thing to do in this situation since the user can't really figure out the ports anyways.

I think you should extract this port-checking-and-determination to a separate function that can be tested.  Something like
```
grpcPort, httpPort := evaluateOptions(opts)
if grpcPort < 0 && httpPort < 0 {
    logrus.Debug("skaffold API not starting as it's not requested")
    return emptyCallback, nil
}
```

Port 0 may make sense for users who don't care about the port but are streaming events for debugging purposes.

Reply:
Now that --enable-rpc is gone, here are the possible scenarios:


input1 | input2 | http port | grpc port | 
--|--|--|--|
-rpc-port=unset  | -http-port=unset | not started | not started
-rpc-port=X  | -http-port=unset | not started | X
-rpc-port=X  | -http-port=Y | Y | X
-rpc-port=unset  | -http-port=Y | Y | bind to a random port with :0

We also don't need 50051 & 50052 defaults anymore (CC does not rely on these).

We can extract to a fn and try to test it but the check is basically:

    if both grpc and http port opts are unset {
       return
    }

@briandealwis thoughts?

Reply:
This table is great it needs `--enable-rpc` too: we need to support `--enable-rpc` until we remove `--enable-rpc` after the deprecation period.

> We also don't need 50051 & 50052 defaults anymore (CC does not rely on these).

But users may.  CC is not the only user of the gRPC + REST APIs.

Reply:
> This table is great it needs `--enable-rpc` too: we need to support `--enable-rpc` until we remove `--enable-rpc` after the deprecation period.

Currently, this is what I had in mind without incorporating the default ports

--enable-rpc | --rpc-port | --http-port | http port | grpc port | 
|--|--|--|--|--|
| false  | *  | *  | not started | not started
| unset | -rpc-port=unset  | -http-port=unset | not started | not started <sup>1</sup>
| unset | -rpc-port=unset  | -http-port=unset | not started | not started <sup>1</sup>
| true | -rpc-port=unset  | -http-port=unset | bind random (:0)  <sup>2</sup> | bind random (:0) <sup>2</sup>
| true or unset | -rpc-port=X  | -http-port=unset | not started | X
| true or unset | -rpc-port=X  | -http-port=Y | Y | X
| true or unset | -rpc-port=unset  | -http-port=Y | Y | bind random (:0)

<sup>1</sup>: This is breaking change for users who were relying on skaffold dev/debug commands starting the server implicitly by default.

<sup>2</sup>: This is breaking change for users who were relying on default ports.

> But users may. CC is not the only user of the gRPC + REST APIs.

Given the complexity of the table above, I'm not sure if we can roll it forward without breaking anyone. Let me know if you see a way. My plan was to roll this forward in lockstep with CC team.

Reply:
Refined the table above, make sure to refer to the one on GitHub UI.

Reply:
I don't know that there's a way to make these changes without breaking an existing flow, and given that I think we should choose the flow that we think will affect the least users, which to me feels like providing `--enable-rpc=true` without specifying either port explicitly. users who are expecting the server to be available on `50051/50052` will be broken, but can introspect on the logs for the port we've selected, and can also begin passing the port explicitly to get their existing behavior back. we can even issue a specific deprecation notice/suggestion for this specific flow.

my intuition is that this will affect a small subset of users and the fix is relatively painless. @briandealwis WDYT

Reply:
This is my suggestion

--enable-rpc | --rpc-port | --http-port | http port | grpc port | 
|--|--|--|--|--|
| false  | *  | *  | not started | not started
| true | unset  | unset | ~bind random (:0)~ **port 50052, fail if unavailable** | ~bind random (:0)~ **port 50051, fail if unavailable**
| true | X  | unset | **port 50052, fail if unavailable** | X
| true | unset  | Y | Y | **port 50051, fail if unavailable**
| ~true or~ **unset** | X  | unset | not started | X
| ~true or~ **unset** | X  | Y | Y | X
| ~true or~ **unset** | unset  | Y | Y | bind random (:0)

So the only change is that `--enable-rpc` will fail when either port 50051 or 50052 are unavailable whereas we previously allocating a new nearby port.  The warning text can recommend using `--rpc-port=0` or `--rpc-http-port=0` for a randomly assigned port.

We could even collapse the `--enable-rpc == false` with the `--enable-rpc == unset` cases

Reply:
(I think this agrees with Nick)

Reply:
@briandealwis problem with the changes suggested in rows 3 and 4 is that we are removing "if busy find adjacent port" logic.

So if you start HTTP server on :X, but we default gRPC to port 50051, the second instance of skaffold running simultaneously will fail to launch.

Therefore, I recommend retaining the recommended behavior in rows 3 and 4. 

We can implement your recommendation in row 2 and add a warning message for the case both ports are unset.

Reply:
> problem with the changes suggested in rows 3 and 4 is that we are removing "if busy find adjacent port" logic.

To be clear, this is going to happen if someone's explicitly setting the --enable-rpc (i.e. IDE integration or otherwise) and they're running multiple `skaffold` processes. Cloud Code currently does this (it specifies either --rpc-port or --rpc-http-port), so if they somehow they start a 2nd skaffold instance, we'd fail if we choose the legacy default port and another process was using it.

Reply:
@briandealwis @nkubala I am inclined to keep the initially proposed behavior here

> So the only change is that `--enable-rpc` will fail when either port 50051 or 50052 are unavailable whereas we previously allocating a new nearby port. The warning text can recommend using `--rpc-port=0` or `--rpc-http-port=0` for a randomly assigned port.

This will be a problem for two reasons.

1. if we keep such default ports, they can't be "default values" on the flags, because setting the flag now means enable the API, which we're moving away from
2. it's problematic for Cloud Code implementations such as VSCode, where they choose -rpc-http-port explicitly. if "we choose -rpc-port as 50051 and fail if busy", then multiple skaffold instances ran by CC for VSC will fail, we don't want this during the rollout of this.

The only thing I want to add is: If `--enable-rpc` is specified but no ports whatsoever, print a warning saying skaffold is choosing ports.